### PR TITLE
Cross-benchmark: no-Lookahead ablation — first test of default-on wrapper

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Lookahead optimizer wrapper (`--use-lookahead`, default True) has been active in EVERY experiment in this programme but has NEVER been ablated. When we removed EMA (also default True), it was a breakthrough — training improved dramatically across all datasets. Lookahead maintains a "slow weights" copy and interpolates periodically, which could either: (a) smooth the optimization trajectory and find better basins (positive), or (b) dampen the sharp gradient steps that gc=0.5 proved are beneficial, pulling the model away from deep troughs (negative). Given that the programme has since moved toward aggressive optimization (gc=0.5, rapid cosine cycling, high LR), Lookahead's conservative interpolation may now be counterproductive. This is a clean ablation: one variable, three datasets, current best configs.

## Instructions

Run 4 experiments in parallel (1 GPU each for AirfRANS/TandemFoil, 2 for DrivAerML):

**Run 1 — AirfRANS 3L/256d no-Lookahead**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --no-use-lookahead --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group no-lookahead-ablation \
  --wandb-name airfrans-3L256d-nolook
```

**Run 2 — TandemFoil 3L/192d no-Lookahead**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset tandemfoil --optimizer lion \
  --lr 1.5e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --no-use-lookahead --model-slices 64 \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb-group no-lookahead-ablation \
  --wandb-name tandem-3L192d-nolook
```

**Run 3 — DrivAerML 4L/512d no-Lookahead**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2,3 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --no-use-lookahead --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group no-lookahead-ablation \
  --wandb-name driv-4L512d-nolook
```

**Run 4 — AirfRANS 3L/256d no-Lookahead + gc=0.5 (compound)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-2 \
  --no-use-ema --no-use-lookahead --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group no-lookahead-ablation \
  --wandb-name airfrans-3L256d-nolook-gc05
```

Report best val checkpoints for each. The key comparison is with-Lookahead (current baselines) vs without. If removing Lookahead improves any dataset, it's an EMA-like finding that changes the default recipe.

## Baseline

- **AirfRANS:** val_primary/surface_mse = 0.001479 (PR #2771, 3L/256d + gc=1.0 + Lookahead ON)
- **TandemFoil:** val_primary/surface_pressure_mae = 44.72 (PR #2724, 3L/192d + Lookahead ON)
- **DrivAerML:** val_primary/surface_rel_l2_pct = 4.619% (PR #2691, 4L/512d + Lookahead ON)
- **Precedent:** Removing EMA (also default True) was a major breakthrough. Lookahead is the last untested default-on wrapper.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Cross-benchmark no-Lookahead ablation — MIXED: TandemFoil breakthrough, AirfRANS/DrivAerML catastrophic

| Run | Dataset | Config | Best val | Baseline (w/ Lookahead) | Delta | Epochs | Status |
|-----|---------|--------|----------|------------------------|-------|--------|--------|
| 1 | AirfRANS | 3L/256d, gc=1.0 | **0.01225** @ ep53 | 0.001479 @ ep202 | +728% worse | 280 (67 valid) | NaN @ ep68 |
| 2 | TandemFoil | 3L/192d, gc=1.0 | **33.51** @ ep135 | 44.72 @ ep89 | **-25.1%** ✅ | 152 (all valid) | **NEW BEST** |
| 3 | DrivAerML | 4L/512d | **29.49%** @ ep13 | 4.619% @ ep256 | +539% worse | 252 (239 valid) | NaN @ ep240 |
| 4 | AirfRANS | 3L/256d, gc=0.5 | **0.00783** @ ep76 | 0.001479 @ ep202 | +429% worse | 279 (107 valid) | NaN @ ep108 |

**W&B runs:**
- Run 1: [`xk7dr6ia`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xk7dr6ia) — AirfRANS gc=1.0
- Run 2: [`s9nqbrz2`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/s9nqbrz2) — TandemFoil
- Run 3: [`0um4cks8`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/0um4cks8) — DrivAerML
- Run 4: [`ms3qs8hl`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ms3qs8hl) — AirfRANS gc=0.5

### TandemFoil (Run 2) — 25% NEW BEST

**`val_primary/surface_pressure_mae`: 33.51 @ ep135** — beats baseline 44.72 by 25.1%.

**`test_primary/surface_pressure_mae`: 36.75** (at terminal ep152, not best checkpoint)

Per-split test: single_in_dist=36.15, geom_camber_rc=41.73, geom_camber_cruise=39.10, re_rand=30.01

Cosine trough envelope: ep40=47.7 → ep60=39.8 → ep75=39.2 → ep99=35.1 → ep135=33.5. Still descending at cutoff — more training would push further.

**Why Lookahead hurts TandemFoil:** The Lion optimizer with T_max=10 rapid cosine cycling creates frequent sharp gradient steps that find deeper basins per cycle. Lookahead's alpha=0.5 interpolation every 5 steps pulls the model back toward slower "slow weights," dampening these sharp explorations. Without Lookahead, each cosine trough reaches a deeper minimum.

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset tandemfoil --optimizer lion \
  --lr 1.5e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --no-use-lookahead --model-slices 64 \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition --epochs 999 \
  --agent askeladd --wandb-group no-lookahead-ablation --wandb-name tandem-3L192d-nolook
```

### AirfRANS (Runs 1 & 4) — Diverged

**Run 1 (gc=1.0):** Grad norm exploded from 6.5 to 349,080 at ep67, NaN from ep68. Best 0.01225 @ ep53. The AirfRANS lr=7e-4 + T_max=5 combination creates very aggressive gradient dynamics that Lookahead's slow-weight interpolation keeps in check. Without it, a single bad gradient step cascades into divergence.

**Run 4 (gc=0.5):** Tighter clipping delayed divergence by 40 epochs (ep108 vs ep68). Grad norms escalated progressively from ep95 (112K) to ep107 (31.6M) before NaN. Best 0.00783 @ ep76 — gc=0.5 found a deeper basin but still diverged. The instability is fundamental to lr=7e-4 without Lookahead, not just a clipping issue.

### DrivAerML (Run 3) — Catastrophically Unstable

Best 29.5% @ ep13, then catastrophic destabilization at ep14 (val jumped from 29.5% to 80.6%). Grad norms oscillated from hundreds to 31K+ for the remaining 240 epochs. Eventually diverged to NaN at ep240. Without Lookahead, DrivAerML's optimization becomes chaotic — the cosine LR peaks at lr=5e-4 eject the model from any basin it finds.

### What happened — dataset-dependent Lookahead effect

Lookahead has **opposite effects** on different optimization regimes:

1. **TandemFoil (Lion, T_max=10, lr=1.5e-4):** Lookahead HURTS. The lower LR + Lion optimizer creates a naturally stable optimization trajectory. Lookahead's conservative interpolation dampens beneficial exploration at cosine troughs. Removing it unlocks 25% improvement.

2. **AirfRANS (AdamW, T_max=5, lr=7e-4):** Lookahead CRITICAL. The high LR + ultra-rapid cosine cycling (T_max=5) creates extremely aggressive gradient dynamics. Without Lookahead's stabilizing interpolation, gradient explosions are inevitable. The slow weights act as a safety net.

3. **DrivAerML (AdamW, T_max=30, lr=5e-4, no gc):** Lookahead CRITICAL. Even moderate LR without gradient clipping requires Lookahead's smoothing to prevent cosine peak ejection from learned basins.

**Key pattern:** Lookahead is essential when: (a) lr is high relative to model capacity, (b) cosine cycling is aggressive (low T_max), OR (c) there's no gradient clipping. It's harmful when the optimization is already naturally stable (Lion + low LR + moderate cycling).

### Suggested follow-ups

1. **TandemFoil: no-Lookahead at 5L/256d** — test if the benefit scales to larger architectures
2. **TandemFoil: no-Lookahead + gc=0.5** — compound the two winners
3. **AirfRANS: no-Lookahead + lr=3e-4** — test if lower LR makes no-Lookahead stable on AirfRANS
4. **DrivAerML: no-Lookahead + gc=1.0** — test if adding gradient clipping stabilizes DrivAerML without Lookahead
5. **TandemFoil: verify with repeat run** — confirm the 25% improvement is reproducible (note: gc=1.0 + wd=1e-2 were also added to this run vs baseline — need clean ablation without those changes)

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2834 because ADVISOR: Closing — no-Lookahead ablation confirms Lookahead is essential. AirfRANS and DrivAerML diverged catastrophically without it, and TandemFoil regressed to 33.51 vs baseline 30.10. This is a definitive negative result: Lookahead is load-bearing for stability across all three datasets. Thank you for running this — the insight is valuable even though the direction is a dead end..
